File: Boss.java
package YushanoBoken.Model;

import java.util.Random;

public class Boss extends Character {

    public Boss(BossName name) {
        switch (name){
            case SOLARIAS_IGNIS:
                this.name = "Solarias Ignis";
                this.hp = 40;
                this.attack = 20;
                this.defense = 20;
                this.exp = 25;
                this.abilty = SpecialAbilty.Burning_Pride;
                break;
            case VACORIUS_NEXIS:
                this.name = "Vacorius Nexis";
                this.hp = 60;
                this.attack = 25;
                this.defense = 25;
                this.exp = 30;
                this.abilty = SpecialAbilty.Chains_of_the_Void;
                break;
            case SOLARIAS_ULTIMA:
                this.name = "Solaris Ultima";
                this.hp = 60;
                this.attack = 30;
                this.defense = 30;
                this.exp = 35;
                this.abilty = SpecialAbilty.Unstoppable_Force;
                break;
            case SHIN_GODZILLA:
                this.name = "Shin Godzilla";
                this.hp = 150;
                this.attack = 60;
                this.defense = 60;
                this.exp = 100000;
                this.abilty = SpecialAbilty.Atomic_Ray;
                break;
        }
    }

    protected SpecialAbilty abilty;

    public SpecialAbilty getAbilty() {
        return abilty;
    }

    public boolean isDead(){
        return getHp() <= 0;
    }

    public Item dropLoot(){
        if (isDead() == true){
            ItemsType[] items = ItemsType.values();
            Random rand = new Random();
            int randomIndex = rand.nextInt(items.length);
            return new Item(items[randomIndex]);
        }
        return null;
    }

    public void useSpecialAbility(Hero hero){
       switch (abilty){
           case Burning_Pride:
               int burningPrideAttack = getAttack() * 150 / 100;
               hero.setEffect(StatusEffect.Burn);
               hero.recieveDamage(burningPrideAttack);
               break;
           case Chains_of_the_Void:
                   hero.setDisable(true);
               break;
           case Unstoppable_Force:
               if(getHp() <= 15){
                   this.attack = getAttack() + 6;
                   this.defense = getDefense() + 6;
               }
               break;
           case Atomic_Ray:
               hero.recieveDamage(getAttack() * 2);
               hero.setEffect(StatusEffect.RADIATION);
               hero.setDisable(true);
               break;
       }
    }
}

File: RPG/Boss.cpp
#include "Boss.h"
#include "Hero.h"
#include <random>

Boss::Boss(BossName name, const char* texturePath, const char* soundPath) {
    // Load assets
    if (texturePath && FileExists(texturePath)) {
        bossTexture = LoadTexture(texturePath);
    }
    if (soundPath && FileExists(soundPath)) {
        abilitySound = LoadSound(soundPath);
    }

    InitializeBossStats(name);
}

Boss::~Boss() {
    if (bossTexture.id != 0) UnloadTexture(bossTexture);
    if (abilitySound.frameCount != 0) UnloadSound(abilitySound);
}

void Boss::InitializeBossStats(BossName name) {
    switch (name) {
    case BossName::SOLARIAS_IGNIS:
        SetName("Solarias Ignis");
        SetHp(40);
        SetAttack(20);
        SetDefense(20);
        SetExp(25);
        ability = SpecialAbility::BURNING_PRIDE;
        break;

    case BossName::VACORIUS_NEXIS:
        SetName("Vacorius Nexis");
        SetHp(60);
        SetAttack(25);
        SetDefense(25);
        SetExp(30);
        ability = SpecialAbility::CHAINS_OF_THE_VOID;
        break;

    case BossName::SOLARIAS_ULTIMA:
        SetName("Solaris Ultima");
        SetHp(60);
        SetAttack(30);
        SetDefense(30);
        SetExp(35);
        ability = SpecialAbility::UNSTOPPABLE_FORCE;
        break;

    case BossName::SHIN_GODZILLA:
        SetName("Shin Godzilla");
        SetHp(150);
        SetAttack(60);
        SetDefense(60);
        SetExp(100000);
        ability = SpecialAbility::ATOMIC_RAY;
        break;
    }
}

bool Boss::IsDead() const {
    return GetHp() <= 0;
}

std::unique_ptr<Item> Boss::DropLoot() {
    if (!IsDead()) return nullptr;

    static std::random_device rd;
    static std::mt19937 gen(rd());
    std::uniform_int_distribution<> dist(0, 3);

    ItemsType randomType = static_cast<ItemsType>(dist(gen));
    return std::make_unique<Item>(randomType);
}

void Boss::UseSpecialAbility(Hero& hero) {
    if (abilitySound.frameCount != 0) {
        PlaySound(abilitySound);
    }

    switch (ability) {
    case SpecialAbility::BURNING_PRIDE: {
        int boostedAttack = GetAttack() * 150 / 100;
        hero.SetEffect(StatusEffect::BURN);
        hero.ReceiveDamage(boostedAttack);
        break;
    }

    case SpecialAbility::CHAINS_OF_THE_VOID:
        hero.SetDisabled(true);
        break;

    case SpecialAbility::UNSTOPPABLE_FORCE:
        if (GetHp() <= 15) {
            SetAttack(GetAttack() + 6);
            SetDefense(GetDefense() + 6);
        }
        break;

    case SpecialAbility::ATOMIC_RAY:
        hero.ReceiveDamage(GetAttack() * 2);
        hero.SetEffect(StatusEffect::RADIATION);
        hero.SetDisabled(true);
        break;
    }
}

void Boss::Draw(Vector2 position) const {
    if (bossTexture.id != 0) {
        DrawTextureV(bossTexture, position, WHITE);
    }
    else {
        // Fallback drawing
        DrawCircleV(position, 30, RED);
        DrawText(GetName().c_str(), position.x - 50, position.y - 50, 20, WHITE);
    }
}